### DISCUSSION TITLE
Idea to avoid inflated speeds due to GPX uploads

### DISCUSSION THREAD

**Discussion by Logiqx** [djong1/gp3s#134] [GPS Technology (devices and formats)]
It seems the average guy on the beach does not fully appreciate that exporting GPX files from COROS / Garmin / Suunto can artifically inflate their speeds when posting to GP3S. I don't think everyone necessarily needs to know how / why this happens but I do believe the issue can be mitigated with some simple software tweaks.

In essence, watch manufacturers such as COROS, Garmin and Suunto do not record speeds in their GPX exports so it needs to be derived from the positional data (i.e. latitude / longitude). Due to the imprecision of lat / lon this typically leads to inflated speeds when GPX files are uploaded to GP3S.

The differences are particularly noticable for 2 sec and 10 sec results but all speed categories are affected to some degree. Take this session as an example, showing 2 sec as 42.11 knots:

![image](https://user-images.githubusercontent.com/17449078/212558831-1a7feb3e-98ff-40af-9edd-94d60563609a.png)

The device itself reported the best 2 sec as 38.60 knots, as shown in the COROS app:

![maciej spike - coros](https://user-images.githubusercontent.com/17449078/212558834-912bb39e-8455-46a7-bbc5-dfe3267c8a4b.jpg)

In this example the speed is inflated by 3.5 knots and it is solely due to a GPX file being posted to GP3S. The corresponding FIT file contains the speed calculated by the watch during the session and doesn't have the issue. Whilst it is possible to talk to people one-by-one this can be quite time consuming and tbh it feels like it shouldn't really be necessary.

It occurred to me today that this kind of error can actually be avoided with some relatively simple processing in the GP3S auto-upload backend (and GPSResults, GPSAR, etc). All GPX files contain a "creator" attribute in their `<gpx>` element such as "Coros Wearables" or "Garmin Connect". Here's an example of a GPX file from a COROS watch.

![image](https://user-images.githubusercontent.com/17449078/212559482-7b4c69ac-dc90-4eae-94e9-9de5b204a6cb.png)

The auto-upload (or laptop software) could simply check the "creator" attribute of the GPX file and whether it is a vendor that provides FIT exports. COROS, Garmin and Suunto all provide FIT exports so any attempt to upload a GPX file from those brands could be aborted and the user notified that they must use the FIT file.

Sure, people could manually edit the GPX file to be a different "creator" but I think this is more to do with people who simply don't know that the GPX file isn't appropriate. The auto-upload could also check if speed is present in the GPX and warn the user if it is not present.

One final note is that the estimated speeds present in the FIT files (which typically appear to be Doppler-derived) are almost always superior to the positional data. I have spent a fair bit of time studying data from the Garmin Fenix 3, 5, 6, 7, Vivoactive 3, COROS APEX Pro, VERTIX and VERTIX 2. Out of all those devices it was clear that the speed data in the FIT files is superior ro than speeds derived from positional data in the GPX. The one exception was the Fenix 7 which appeared to have some strange speed filtering.

Of course we don't know for sure the method by which speeds are estimated in every GPS / GNSS chipset (or even added by Garmin, etc). 5 or 6 techniques exist for the estimation of speed from the raw GNSS observables (psueodrange, carrier-phase and Doppler); one is simply position-derived (referred to as non-Doppler by the speedsurfing community) and one is Doppler-derived (referred to as Doppler speed by the speedsurfing community). The other methods include time-differenced pseudorange, carrier-derived-Doppler, time-differenced carrier-phase, all of which are have been published for decades and could potentially be used in modern-day GNSS chipsets. The GNSS chipset may also use a least-squares estimator or Kalman filter but whatever the approach taken, it's clear that the speed is always of higher quality than the positional data in consumer devices.

To summarise, it should be relatively straightforward to ensure users post the appropriate data from their device by simply checking the content of the GPX file. The main manufacturers (COROS, Garmin, Suunto) all provide FIT exports and people really should be using them to avoid inflated speeds. I see several results where I know for a fact the people have posted inflated speeds, gaining several knots which is kinda unfair on other competitors imo.

Thought for the future - if / when manufacturer include speed in their GPX files they could be supported as well. The safety check might therefore check the "creator" and the presence of speed. I wrote a separate article about GPX files and the abundance of issues related to non-conformance with the standard. It turns out that almost no manufacturer generates GPX files correctly, especially GPX 1.1 files - [link](https://logiqx.github.io/gps-wizard/gpx/)

**Comment by djong1** [comment]
Hi,

Thanks for pointing this out. I would like to discuss this a bit more based on our current approach. We have now marked GPX usage as a yellow icon:
Home - gps-speedsurfing<https://gps-speedsurfing.com/default.aspx?mnu=forum&forum=2&val=191866> when using fit file it is a green icon.

[https://cdn.gps-speedsurfing.com/gp3s/cdn/img/gp3-logo-rechthoekig.jpeg]<https://gps-speedsurfing.com/default.aspx?mnu=forum&forum=2&val=191866>
Home - gps-speedsurfing<https://gps-speedsurfing.com/default.aspx?mnu=forum&forum=2&val=191866>
GPS-Speedsurfing, for windsurfers by windsurfers. GPS-Speedsurfing is a worldwide competition for speed windsurfers. Take your GPS, upload your data and your are listed in the ranking
gps-speedsurfing.com
This icon is only visible in lists and not on session detail page, should we add the icon here also?

Greetings Dylan
________________________________
Van: Michael George ***@***.***>
Verzonden: zondag 15 januari 2023 20:08
Aan: djong1/gp3s ***@***.***>
CC: Subscribed ***@***.***>
Onderwerp: [djong1/gp3s] Idea to avoid inflated speeds due to GPX uploads (Discussion #134)


It seems the average guy on the beach does not fully appreciate that exporting GPX files from COROS / Garmin / Suunto can artifically inflate their speeds when posting to GP3S. I don't think everyone necessarily needs to know how / why this happens but I do believe the issue can be mitigated with some simple software tweaks.

In essence, watch manufacturers such as COROS, Garmin and Suunto do not record speeds in their GPX exports so it needs to be derived from the positional data (i.e. latitude / longitude). Due to the imprecision of lat / lon this typically leads to inflated speeds when GPX files are uploaded to GP3S.

The differences are particularly noticable for 2 sec and 10 sec results but all speed categories are affected to some degree. Take this session as an example, showing 2 sec as 42.11 knots:

[image]<https://user-images.githubusercontent.com/17449078/212558831-1a7feb3e-98ff-40af-9edd-94d60563609a.png>

The device itself reported the best 2 sec as 38.60 knots, as shown in the COROS app:

[maciej spike - coros]<https://user-images.githubusercontent.com/17449078/212558834-912bb39e-8455-46a7-bbc5-dfe3267c8a4b.jpg>

In this example the speed is inflated by 3.5 knots and it is solely due to a GPX file being posted to GP3S. The corresponding FIT file contains the speed calculated by the watch during the session and doesn't have the issue. Whilst it is possible to talk to people one-by-one this can be quite time consuming and tbh it feels like it shouldn't really be necessary.

It occurred to me today that this kind of error can actually be avoided with some relatively simple processing in the GP3S auto-upload backend (and GPSResults, GPSAR, etc). All GPX files contain a "creator" attribute in their <gpx> element such as "Coros Wearables" or "Garmin Connect". Here's an example of a GPX file from a COROS watch.

[image]<https://user-images.githubusercontent.com/17449078/212559482-7b4c69ac-dc90-4eae-94e9-9de5b204a6cb.png>

The auto-upload (or laptop software) could simply check the "creator" attribute of the GPX file and whether it is a vendor that provides FIT exports. COROS, Garmin and Suunto all provide FIT exports so any attempt to upload a GPX file from those brands could be aborted and the user notified that they must use the FIT file.

Sure, people could manually edit the GPX file to be a different "creator" but I think this is more to do with people who simply don't know that the GPX file isn't appropriate. The auto-upload could also check if speed is present in the GPX and warn the user if it is not present.

One final note is that the estimated speeds present in the FIT files (which typically appear to be Doppler-derived) are almost always superior to the positional data. I have spent a fair bit of time studying data from the Garmin Fenix 3, 5, 6, 7, Vivoactive 3, COROS APEX Pro, VERTIX and VERTIX 2. Out of all those devices it was clear that the speed data in the FIT files is superior ro than speeds derived from positional data in the GPX. The one exception was the Fenix 7 which appeared to have some strange speed filtering.

Of course we don't know for sure the method by which speeds are estimated in every GPS / GNSS chipset (or even added by Garmin, etc). 5 or 6 techniques exist for the estimation of speed from the raw GNSS observable; one is simply position-derived (refer to as non-Doppler by the speedsurfing community) and one is Doppler-derived (referred to as Doppler speed). The other methods include time-differenced pseudorange, carrier-derived-Doppler, time-differenced carrier-phase, all of which are have been published for decades and potentially used by some nodern-day GNSS chipsets. The GNSS chipset may also use a least-squares estimator or Kalman filter but whatever the approach taken, it's clear that the speed is always of higher quality than the positional data in consumer devices.

To summarise, it should be relatively straightforward to ensure users post the appropriate data from their device by simply checking the content of the GPX file. The main manufacturers (COROS, Garmin, Suunto) all provide FIT exports and people really should be using them to avoid inflated speeds. I see several results where I know for a fact the people have posted inflated speeds, gaining several knots which is kinda unfair on other competitors imo.

Thought for the future - if / when manufacturer include speed in their GPX files they could be supported as well. The safety check might therefore check the "creator" and the presence of speed. I wrote a separate article about GPX files and the abundance of issues related to non-conformance with the standard. It turns out that almost no manufacturer generates GPX files correctly, especially GPX 1.1 files - link<https://logiqx.github.io/gps-wizard/gpx/>

**Comment by Logiqx** [comment]
That's a good idea. I think it would be good to have the green / yellow on the session detail and it would go well next to the GPS type:

![image](https://user-images.githubusercontent.com/17449078/212735529-548d57cc-c0bd-442c-8a8f-7d9b36e6db2f.png)

The satellite icon is very useful in the rankings pages:

![image](https://user-images.githubusercontent.com/17449078/212728256-cda02788-7587-499f-89ee-f6e6df0b3a4c.png)

You will recall I once suggested adding it to the session lists. Aside from left / right alignment of the name the screens are very similar so there is probably room for the satellite icon:

![image](https://user-images.githubusercontent.com/17449078/212728500-85c31377-99ca-4f85-9fb7-e682a957855c.png)

The satellite icon would be particularly useful in user session lists as well. There is more room than the session lists because name is not shown:

![image](https://user-images.githubusercontent.com/17449078/212728839-91cb2d85-5b0c-478d-a8b1-788aed2e3a44.png)

The event sessions page is quite busy so space may be an issue for the satellite icon but not out of the question?

![image](https://user-images.githubusercontent.com/17449078/212730606-51d9d77a-adf9-4ae5-8969-aa6b69c6e6d1.png)

Whilst writing this comment, I started to wonder whether it might be worth showing the speeds from GPX sessions in yellow, rather than white. If everyone posts Doppler then rankings and session lists won't look any different but any GPX sessions would stand out:

![image](https://user-images.githubusercontent.com/17449078/212732491-c1d9dcee-8544-4ad1-824f-221cee77b4ea.png)

This idea of yellow text could be applied to pretty much any list / ranking (including "ranking of the day") where a GPX speed appears. It has the benefit of not requiring extra space for the satellite icon. It should be pretty easy to implement via an additional "gpx" class on the `<td>` elements and suitable CSS.

So to summarise:

- I think the green / yellow satellite icon on the session page would be a good idea and is would look good next to the GPS type
- Some / all of the session lists would also benefit from a green / ywllow satellite icon but I recognise you don't want over-crowding
- Yellow text could (maybe) replace the satellite icon on all rankings / session lists. The icon would still appear on the session details page.

I think there is also some mileage in pre-screening GPX uploads, prior to processing with the GPSResults API. It should just be a matter of checking the "creator" attribute in the XML. COROS / Garmin / Suunto files are easy to identify and the user could be instructed to upload the FIT. This would be a simple way to stop people unwittingly (or intentionally) posting inflated speeds due to GPX. It obviously requires a slightly different workflow but I suspect it wouldn't be particularly challenging?

**Comment by djong1** [comment]
Thanks, that is super helpful feedback. I will implement your line of thinking for the icons and text colors.  People that upload gpx while fit is available will receive in the future each time a personal message via email and in the message board.

Verzonden vanuit Outlook voor iOS<https://aka.ms/o0ukef>
________________________________
Van: Michael George ***@***.***>
Verzonden: Monday, January 16, 2023 6:38:39 PM
Aan: djong1/gp3s ***@***.***>
CC: Dylan de Jong ***@***.***>; Comment ***@***.***>
Onderwerp: Re: [djong1/gp3s] Idea to avoid inflated speeds due to GPX uploads (Discussion #134)


That's a good idea. I think it would be good to have the green / yellow on the session detail and it would go well next to the GPS type:

[image]<https://user-images.githubusercontent.com/17449078/212735529-548d57cc-c0bd-442c-8a8f-7d9b36e6db2f.png>

The satellite icon is very useful in the rankings pages:

[image]<https://user-images.githubusercontent.com/17449078/212728256-cda02788-7587-499f-89ee-f6e6df0b3a4c.png>

You will recall I once suggested adding it to the session lists. Aside from left / right alignment of the name the screens are very similar so there is probably room for the satellite icon:

[image]<https://user-images.githubusercontent.com/17449078/212728500-85c31377-99ca-4f85-9fb7-e682a957855c.png>

The satellite icon would be particularly useful in user session lists as well. There is more room than the session lists because name is not shown:

[image]<https://user-images.githubusercontent.com/17449078/212728839-91cb2d85-5b0c-478d-a8b1-788aed2e3a44.png>

The event sessions page is quite busy so space may be an issue for the satellite icon but not out of the question?

[image]<https://user-images.githubusercontent.com/17449078/212730606-51d9d77a-adf9-4ae5-8969-aa6b69c6e6d1.png>

Whilst writing this comment, I started to wonder whether it might be worth showing the speeds from GPX sessions in yellow, rather than white. If everyone posts Doppler then rankings and session lists won't look any different but any GPX sessions would stand out:

[image]<https://user-images.githubusercontent.com/17449078/212732491-c1d9dcee-8544-4ad1-824f-221cee77b4ea.png>

This idea of yellow text could be applied to pretty much any list / ranking (including "ranking of the day") where a GPX speed appears. It has the benefit of not requiring extra space for the satellite icon. It should be pretty easy to implement via an additional "gpx" class on the <td> elements and suitable CSS.

So to summarise:

  *   I think the green / yellow satellite icon on the session page would be a good idea and is would look good next to the GPS type
  *   Some / all of the session lists would also benefit from a green / ywllow satellite icon but I recognise you don't want over-crowding
  *   Yellow text could (maybe) replace the satellite icon on all rankings / session lists. The icon would still appear on the session details page.

I think there is also some mileage in pre-screening GPX uploads, prior to processing with the GPSResults API. It should just be a matter of checking the "creator" attribute in the XML. COROS / Garmin / Suunto files are easy to identify and the user could be instructed to upload the FIT. This would be a simple way to stop people unwittingly (or intentionally) posting inflated speeds due to GPX. It obviously requires a slightly different workflow but I suspect it wouldn't be particularly challenging?

**Comment by Logiqx** [comment]
@djong1 - Good stuff. Something to look forward too in a future update. :)

I have also noticed that some sessions do not show a GPS type. I've always assumed it gets hidden when a COROS user uploads their session as a GPX, perhaps for other reasons as well? Once the satellite icon is shown with the session details, GPS type can be unhidden for such sessions and shown with a yellow satellite icon?

**Reply by jhdebruin** [reply]
_"Yellow text could (maybe) replace the satellite icon on all rankings / session lists. The icon would still appear on the session details page"_


From a UI perspective i think we should avoid to use the yellow text in views, this will break the UI style to much

**Reply by Logiqx** [reply]
That's a good shout. Maybe it would be better to use the same grey colour as person names / spot names?

That would make the speeds less prominent in the various lists but be consistent with the palette already in use.

**Comment by Logiqx** [comment]
Further to the comment by @jhdebruin, I think using the same grey as person / spot names would work nicely.

Here's a quick illustration (although not quite the same font) using the foil rankings:

![image](https://user-images.githubusercontent.com/17449078/212831302-4abec4f8-5ad7-4153-898f-70b2ecf6a2fc.png)

It's a shame the wing rankings have an especially high number of GPX posts because most wingers report on their top speeds.

Bruno's 34.40 knots was from a Suunto GPX file during WSW. His GT-31 gave 32.32 knots and iirc the Suunto itself said about the same. That's 2 knots extra from the GPX file. Who knows what Christian's top speed was but it probably wasn't 36.8 knots with a 27.47 knots 10 sec. Luca is probably the fastest in that list, rather than third.

I think showing such speeds in grey is useful when looking at GPX speeds. It's quite subtle, yet still quite apparent. Ultimately the colour can be changed via one line of CSS but I think the grey is probably a good choice?

**Reply by jhdebruin** [reply]
Quick answer, the "grey" option is much better indeed @djong1

**Comment by Logiqx** [comment]
@djong1 would it also be possible to include the same indicator in API results; eventranking, eventsessions, etc?

**Comment by djong1** [comment]
First update will be deployed soon:

<img width="656" alt="voorbeeld" src="https://user-images.githubusercontent.com/25177991/215352972-35ffe271-e7b0-4801-9d15-41fa60a1d518.png">

when a user uploads fit, it will be green and when it is gpx it will be yellow

<img width="655" alt="image" src="https://user-images.githubusercontent.com/25177991/215353116-08187383-eb3a-474f-94a9-21fdbebd1676.png">

**Reply by Logiqx** [reply]
Nice!

**Comment by Logiqx** [comment]
The new icon implemented at the end of January is nice.

I've noticed that some sessions do not list a GPS and thus no icon either. Do you know why?

A few examples:

- https://www.gps-foilsurfing.com/default.aspx?mnu=user&val=378433&uid=5880&spotid=761
- https://www.gps-speedsurfing.com/default.aspx?mnu=user&val=254849&uid=4824&spotid=1642
- https://www.gps-foilsurfing.com/default.aspx?mnu=user&val=401316&uid=26661&spotid=262

It may happen (but not exclusively) when approved devices such as COROS files are uploaded as GPX?

**Reply by djong1** [reply]
If the brand was unknown we hide this one, I have changed it.

**Comment by Logiqx** [comment]
On another note it would also be nice if the idea about speeds being listed in grey could also be implemented.

Re-posting screenshot from above:

![image](https://user-images.githubusercontent.com/17449078/230067556-9d4ea33a-ac2a-49f7-9136-e6e9495f63e9.png)

**Reply by jhdebruin** [reply]
Thnx @Logiqx , think we have to discuss this in the team once again, maybe we have some other options as well...

**Reply by Logiqx** [reply]
Cheers. If you want to bounce any ideas off me, feel free to drop me a PM.
